### DISCUSSION TITLE
Calculating flow through a boundary

### DISCUSSION THREAD

**Discussion by nlosacco** [idaholab/moose#26745] [Q&A Modules: Porous Flow]
For my problem I have to impose a time-varying porepressure at a boundary through a `DirichletBC`. This seems to be the best option for my case, compared to Neumann bc's like e.g. `PorousFlowSink`. Now I need the total flow through the boundary but I could not find a straightforward way to obtain it. When working with the simple Diffusion Kernel I could use a `SideDiffusiveFluxIntegral` but I have no idea if I can employ it with PorousFlow (e.g. how do I define the correct `diffusivity` given that the permeability might be anisotropic and changing with time?).
Maybe I can calculate the flow by first using `PorousFlowDarcyVelocityComponent`, but again I have no clue on how to do that.

**Comment by GiudGiud** [comment]
> how do I define the correct diffusivity given that the permeability might be anisotropic and changing with time?

In that kind of situation, I think the answer is to write your own version of `SideDiffusiveFluxIntegral` that has the right diffusivity
There is an advective flux postprocessor as well, but I am not sure it applies to porous flow.

**Reply by nlosacco** [reply]
Wow, that's gonna be hard... I can hardly read the C++ code. I've even found an issue raised by @WilkAndy in 2018 and apparently not solved yet: https://github.com/idaholab/moose/issues/11376
I do not think I am able to play with the code, maybe I can find some AuxKernels, Postprocessors, etc. that can come to aid

**Reply by GiudGiud** [reply]
@cpgr any recommendation?

modifying the C++ is kind of expected when no one has done what you want before

**Reply by cpgr** [reply]
The correct thing to do is probably to make a post processor that does this, but in the mean time, you might be able to approximate it using a combination of SideIntegralVariablePostprocessor and a product of PorousFlowDarcyVelocityComponent AuxVariable and an AuxVariable for density (using PorousFlowPropertyAux)? Or maybe a SideAdvectiveFluxIntegral using a density AuxVariable and the velocity from PorousFlowDarcyVelocityComponent?

**Reply by nlosacco** [reply]
I've found this: https://mooseframework.inl.gov/source/postprocessors/BoundaryFluxPostprocessor.html
that seems to be helpful

but it seems not to work (the corresponding TestBoundaryFlux UserObject is not defined)

**Reply by GiudGiud** [reply]
Hello

No that one is not helpful, it s an object for the RDG module that is coded very differently from porousFlow

**Reply by nlosacco** [reply]
I'm having a headache after spending one day trying to build my own postprocessor.
I know it is cumbersome but I've done:
get velocity components with `PorousFlowDarcyVelocityComponent` stored in `AuxVariables` with `type = MONOMIAL`
get an elemental vector variable with a `ParsedVectorAux` stored as `MONOMIAL_VEC`
use the latter in my own implementation of `SideDiffusiveFluxIntegral`:
```
template <bool is_ad>
Real
SideDiffusiveFluxIntegralNTempl<is_ad>::computeQpIntegral()
{
  return _v * _normals[_qp];
}
```
`_v` is the elemental vector variable (type `const VectorVariableValue`) and _normals[_qp] is of type `const libMesh::Point`
Now when building the application I get:
```
error: no match for 'operator*' (operand types are 'const VectorVariableValue' {aka 'const MooseArray<libMesh::VectorValue<double> >'} and 'const libMesh::Point')
```
I understand the * operator is overloaded somewhere for types `VariableGradient` and `Point`, but I do not know where to look at due to my ignorance of C++. Can you give me a hint, please?

**Reply by GiudGiud** [reply]
yeah i dont know why.

Just do it manually for now
```
return _v(0) * _normals[_qp](0) + _v(1) * _normals[_qp](1) + _v(2) * _normals[_qp](2);
```

**Reply by nlosacco** [reply]
ok, I found out by looking at libMesh documentation that I have to swap the operands:
`return _normals[_qp] * _v[_qp]` compiles.
By the way, is it `_v` or `_v[_qp]` ?
Now build fails with 
`main.x86_64-conda-linux-gnu.opt.lo' is not a valid libtool object`

your suggestion yields:

```
error: no match for call to '(const VectorVariableValue {aka const MooseArray<libMesh::VectorValue<double> >}) (int)'
   38 | return _v(0) * _normals[_qp](0) + _v(1) * _normals[_qp](1) + _v(2) * _normals[_qp](2);

```

**Reply by GiudGiud** [reply]
must be `_v[0]` then

that is unrelated I think you need to clean up your repo, make sure the right compilers are in use

**Comment by Traiwit** [comment]
```
#include "AdvectiveFluxAuxBE.h"
#include "Assembly.h"

registerMooseObject("moose_behApp", AdvectiveFluxAuxBE);

InputParameters
AdvectiveFluxAuxBE::validParams()
{
  InputParameters params = AuxKernel::validParams();
  MooseEnum component("x y z normal");
  params.addClassDescription("Compute components of flux vector for advection problems "
                             "$(\\vec{J} = \\vec{v} \\cdot \\vec{n})$.");
  params.addCoupledVar("vx", "The x component of velocity");
  params.addCoupledVar("vy", "The y component of velocity");
  params.addCoupledVar("vz", "The z component of velocity");
  params.addParam<MooseEnum>("component", component, "The desired component of flux."); // Added this line
  return params;
}

AdvectiveFluxAuxBE::AdvectiveFluxAuxBE(const InputParameters & parameters)
  : AuxKernel(parameters),
    _component(getParam<MooseEnum>("component")),
    _vx(coupledValue("vx")),
    _vy(coupledValue("vy")),
    _vz(coupledValue("vz")),
    _normals(_assembly.normals()),
    _use_normal(getParam<MooseEnum>("component") == "normal")
{
  if (_use_normal && !isParamValid("boundary"))
    paramError("boundary", "A boundary must be provided if using the normal component!");
}

Real
AdvectiveFluxAuxBE::computeValue()
{
  // RealVectorValue velocity(_v[_qp]);
  return (_vx[_qp] * _normals[_qp](0)) + (_vy[_qp] * _normals[_qp](1)) + ((_vz[_qp]) * _normals[_qp](2)) ;
}
```

I've made this a while back, not sure if it's what you want, the vx vy vz are from `PorousFlowDarcyVelocityComponent`

**Reply by nlosacco** [reply]
Hi @Traiwit , tanks so much for your reply.

I deemed it somewhat more logical to use directly the Darcy velocity vector in the calculations 

> get velocity components with `PorousFlowDarcyVelocityComponent` stored in `AuxVariables` with `type = MONOMIAL`
> get an elemental vector variable with a `ParsedVectorAux` stored as `MONOMIAL_VEC`

please find attached the files I used for my postprocessor (called `SideDiffusiveFluxIntegralN` which I should obviously change) I copied the SideDiffusiveFluxIntegral.C and base class files and got ride of the Finite Volume bit. Please find them attached.

I've noted that a `PorousFlowDarcyVelocityMaterial` exists that computes said velocity vectors. Why is it only implemented at quadrature points `at_nodes = false`)? Is it possible to calculate the velocity at the nodes as for other `PorousFlow` materials?

[SideDiffusiveFluxIntegralN.C.txt](https://github.com/idaholab/moose/files/14269960/SideDiffusiveFluxIntegralN.C.txt)
[SideDiffusiveFluxIntegralN.h.txt](https://github.com/idaholab/moose/files/14269962/SideDiffusiveFluxIntegralN.h.txt)
[SideIntegralVariablePostprocessorN.C.txt](https://github.com/idaholab/moose/files/14269963/SideIntegralVariablePostprocessorN.C.txt)
[SideIntegralVariablePostprocessorN.h.txt](https://github.com/idaholab/moose/files/14269964/SideIntegralVariablePostprocessorN.h.txt)

**Reply by GiudGiud** [reply]
Velocity is computed with
```
    _darcy_velocity[_qp][ph] =
        -(_permeability[_qp] * (_grad_p[_qp][ph] - _fluid_density[_qp][ph] * _gravity) *
          _relative_permeability[_qp][ph] / _fluid_viscosity[_qp][ph]);
```
the problem is the gradient of pressure here. If pressure is a lagrange variable, its gradient is not continuous on nodes. You can do some type of gradient averaging / reconstruction but it might not help for getting conservation
